src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,671 @@
+	ret = lib_manager_allocate_module_instance(module_id, IPC4_INST_ID(ipc_config->id),
+						   base_cfg->is_pages, mod);
+	if (ret < 0) {
+		tr_err(&lib_manager_tr, "lib_manager_allocate_module() failed: %d", ret);
+		return 0;
+	}
+	return mod->entry_point;
+}
+
+int lib_manager_free_module(const struct comp_driver *drv,
+			    struct comp_ipc_config *ipc_config)
+{
+	struct sof_man_fw_desc *desc;
+	struct sof_man_module *mod;
+	uint32_t module_id = IPC4_MOD_ID(ipc_config->id);
+	uint32_t entry_index = LIB_MANAGER_GET_MODULE_INDEX(module_id);
+	int ret;
+
+	tr_dbg(&lib_manager_tr, "lib_manager_free_module() mod_id: 0x%x", ipc_config->id);
+
+	desc = lib_manager_get_library_module_desc(module_id);
+	mod = (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(entry_index));
+
+	ret = lib_manager_unload_module(module_id, mod, desc);
+	if (ret < 0)
+		return ret;
+
+	ret = lib_manager_free_module_instance(module_id, IPC4_INST_ID(ipc_config->id), mod);
+	if (ret < 0) {
+		tr_err(&lib_manager_tr, "lib_manager_allocate_module() failed: %d", ret);
+		return ret;
+	}
+	return 0;
+}
+
+void lib_manager_init(void)
+{
+	struct sof *sof = sof_get();
+
+	if (!sof->ext_library)
+		sof->ext_library = &loader_ext_lib;
+}
+
+struct sof_man_fw_desc *lib_manager_get_library_module_desc(int module_id)
+{
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	struct ext_library *_ext_lib = ext_lib_get();
+	uint8_t *buffptr = (uint8_t *)_ext_lib->desc[lib_id];
+
+	return (struct sof_man_fw_desc *)(buffptr + SOF_MAN_ELF_TEXT_OFFSET);
+}
+
+void lib_manager_update_sof_ctx(struct sof_man_fw_desc *desc, uint32_t lib_id)
+{
+	struct ext_library *_ext_lib = ext_lib_get();
+
+	_ext_lib->desc[lib_id] = desc;
+	/* TODO: maybe need to call here dcache_writeback here? */
+}
+
+int lib_manager_register_module(struct sof_man_fw_desc *desc, int module_id)
+{
+	/* allocate new  comp_driver_info */
+	struct comp_driver_info *new_drv_info;
+	struct comp_driver *drv;
+	struct sof_man_module *mod;
+	int ret;
+
+	uint32_t align = 4;
+	uint32_t entry_index = LIB_MANAGER_GET_MODULE_INDEX(module_id);
+
+	new_drv_info = (struct comp_driver_info *)rballoc_align(0, SOF_MEM_CAPS_RAM,
+					sizeof(struct comp_driver_info), align);
+
+	if (!new_drv_info) {
+		tr_err(&lib_manager_tr, "lib_manager_register_module(): alloc failed");
+		ret = -ENOMEM;
+		goto cleanup;
+	}
+
+	drv = (struct comp_driver *)rballoc_align(0, SOF_MEM_CAPS_RAM,
+			sizeof(struct comp_driver), align);
+	if (!drv) {
+		tr_err(&lib_manager_tr, "lib_manager_register_module(): alloc failed");
+		ret = -ENOMEM;
+		goto cleanup;
+	}
+	memset(drv, 0, sizeof(struct comp_driver));
+
+	/* Fill the new_drv_info structure with already known parameters */
+	/* Check already registered components */
+	mod = (struct sof_man_module *)((uint8_t *)desc + SOF_MAN_MODULE_OFFSET(entry_index));
+	struct sof_uuid *uid = (struct sof_uuid *)&mod->uuid[0];
+
+#if CONFIG_INTEL_MODULES
+	/* Declare new dynamic module with Module Adapter */
```

Comment:
why is this under `CONFIG_INTEL_MODULES`. It doesn't really have any intel specific stuff.

Reply:
Not exactly, library manager uses iadk_modules interface which is dependent on intel modules. That is why we used tath